custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/checkstyle/ThingHandlerSetLabelCheck.java

```diff
@@ -0,0 +1,212 @@
+import com.puppycrawl.tools.checkstyle.api.FileText;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.CheckUtils;
+
+import antlr.RecognitionException;
+import antlr.TokenStreamException;
+
+/**
+ * Checks the code and generates a warning if a class extends BaseThingHandler
+ * and the instance method setLabel of the Thing interface is used.
+ * The label have to be set only by the user.
+ *
+ * @author Tanya Georgieva - Initial Contribution
+ */
+public class ThingHandlerSetLabelCheck extends AbstractStaticCheck {
+
+    private static final String WARNING_MESSAGE = "Do not use the setLabel of the thing in the ThingHandler";
+    private static final String BASE_THING_HANDLER = "BaseThingHandler";
+    private static final String SEARCHED_PACKAGE = "org.eclipse.smarthome.core.thing.binding." + BASE_THING_HANDLER;
+    private static final String SET_LABEL = "setLabel";
+    private static final String THING = "thing";
+    private static final String UNKNOWN = "UNKNOWN";
+
+    private final Log logger = LogFactory.getLog(ThingHandlerSetLabelCheck.class);
+
+    private String fullJavaExtension = "\\." + JAVA_EXTENSION;
+
+    private HashMap<File, DetailAST> files;
+    private HashMap<File, String> derivedAndBaseClasses;
+    private HashMap<File, DetailAST> classesExtendingBaseThingHandler;
+
+    public ThingHandlerSetLabelCheck() {
+        setFileExtensions(JAVA_EXTENSION);
+    }
+
+    @Override
+    public void init() {
+        files = new HashMap<>();
+        derivedAndBaseClasses = new HashMap<>();
+        classesExtendingBaseThingHandler = new HashMap<>();
+    }
+
+    @Override
+    protected void processFiltered(File file, FileText fileText) throws CheckstyleException {
+        FileContents fileContents = new FileContents(fileText);
+        DetailAST rootAST = null;
+
+        try {
+            rootAST = TreeWalker.parse(fileContents);
+        } catch (RecognitionException | TokenStreamException e) {
+            logger.error("Parsing file not successful");
+            e.printStackTrace();
+        }
+        files.put(file, rootAST);
+    }
+
+    @Override
+    public void finishProcessing() {
+        findDerivedAndBaseClasses();
+        findClassesExtendingBaseThingHandler();
+
+        classesExtendingBaseThingHandler.forEach((file, rootAST) -> {
+            String fileName = file.getName();
+            Path filePath = file.toPath();
+            checkSetLabelMethodCall(rootAST, fileName, filePath);
+        });
+    }
+
+    // Pre-order traversal - starting from the root first
+    private void checkSetLabelMethodCall(DetailAST nodeAST, String fileName, Path filePath) {
+        if (nodeAST == null) {
+            return;
+        }
+
+        if (THING.equals(nodeAST.getText())) {
+            // thing.setLabel(...)
+            boolean isThingFieldSetLabelMethodCalled = nodeAST.getNextSibling() != null
```

Comment:
If we have a custom thing which we call thing and call `setLabel` on that will that be allowed or report a false positive?

Reply:
As I start searching first for the correct BTH in the classes that extends it directly (after that I am searching for the classes with indirect inheritance) if such classes are found I continue with the thing and setLabel method search. And again if any,  I rely on that there won't be any other instance called exactly "thing" or method "setLabel" in the current class. But yes, if there is such custom thing with the same method name it will be reported.